### DISCUSSION TITLE
migration performance

### DISCUSSION THREAD

**Discussion by rdlaitila** [geldata/gel#3395] [General]
Having issues with a somewhat large concept model where initial migration is taking ~25 minutes to create, and receiving timeouts when attempting to apply them.

edgedb cli: 1.0.0-rc.6+4b9d831
edgedb instance: 1.0-rc.4+1850eff
os: macosx

I can provide some rough counts on the schema:

| schema | count |
| --- | --- |
| abstract types | 4 |
| concrete types | 46 |
| scalar types | 4 |
| total links | 51  |
| total multi links | 5 |
| total propertes | 35 |

When creating the initial migration:

```shell
% time edgedb migration create --non-interactive
edgedb migration create --non-interactive  0.04s user 0.05s system 0% cpu 25:02.23 total
```

When attempting to apply the migration:

```shell
time edgedb migration apply                   
edgedb error: ClientConnectionEosError
edgedb migration apply  0.02s user 0.02s system 0% cpu 11.518 total
```

From the logs I was able to find an entry that seems related:

```shell
INFO 17787 2022-02-02T16:03:12.159 edb.server: Backend connections to database edgedb: 1 were established in at least the last 1.0 seconds.
CRITICAL 17799 2022-02-02T16:03:28.200 postgres: terminating connection due to idle-in-transaction timeout
INFO 17787 2022-02-02T16:03:28.201 edb.server: Connection discarded to backend database: edgedb
```

I can likely split the model and create smaller migrations. However I found the migration creation times quite long and the client timeouts concerning. Let me know if there is any other info the team could find useful.

**Comment by elprans** [comment]
What happens when you do the migration in REPL?

```
start migration to {
  <your schema>
};
populate migration;
commit migration;
```

**Reply by rdlaitila** [reply]
Tried it once and got:

```shell
edgedb> start migration to {
  <my schema>
};
OK: START MIGRATION
edgedb[tx]> populate migration;
edgedb error: connection closed by server
Reconnecting...
edgedb> populate migration;
error: QueryError: cannot execute POPULATE MIGRATION outside of a migration block
```

tried it once more and it seems to work:

```shell
edgedb> start migration to {
  <my schema>
};
OK: START MIGRATION
edgedb[tx]> populate migration;
OK: POPULATE MIGRATION
edgedb[tx]> commit migration;
OK: COMMIT MIGRATION
``` 

Still seeing some instances of "terminating connection due to idle-in-transaction timeout" in the logs but per your other comment you may have found the culprit there.

**Comment by elprans** [comment]
The idle-in-transaction timeout should be fixed by https://github.com/edgedb/edgedb-cli/pull/648

**Comment by rdlaitila** [comment]
for anyone hitting this issue and until a fix is released I used the following workaround for the `ClientConnectionEosError` error:

```shell
edgedb> configure instance set session_idle_transaction_timeout := <duration>'0';
```
